src/common/sv2_noise.h

```diff
@@ -0,0 +1,295 @@
+
+    uint256 GetHashOutput();
+
+    /* For testing */
+    void LogChainingKey();
+    std::string GetChainingKey();
+
+private:
+    uint8_t m_chaining_key[Sv2CipherState::HASHLEN];
+    uint256 m_hash_output = uint256(PROTOCOL_NAME_DOUBLE_HASH);
+    Sv2CipherState m_cipher_state;
+
+    void HKDF2(const Span<const std::byte> input_key_material, uint8_t out0[Sv2CipherState::HASHLEN], uint8_t out1[Sv2CipherState::HASHLEN]);
+};
+
+/*
+ * A HandshakeState object contains a SymmetricState plus DH variables (s, e, rs, re)
+ * and a variable representing the handshake pattern. During the handshake phase
+ * each party has a single HandshakeState, which can be deleted once the handshake
+ * is finished.
+ *
+ * See chapter "5. Processing rules" of the Noise paper.
+ */
+
+class Sv2HandshakeState
+{
+public:
+    static constexpr size_t ELLSWIFT_PUB_KEY_SIZE{64};
+    static constexpr size_t ECDH_OUTPUT_SIZE{32};
+
+    static constexpr size_t HANDSHAKE_STEP2_SIZE = ELLSWIFT_PUB_KEY_SIZE + ELLSWIFT_PUB_KEY_SIZE +
+                                                   Poly1305::TAGLEN + Sv2SignatureNoiseMessage::SIZE + Poly1305::TAGLEN;
+
+    /*
+     * If we are the initiator m_authority_pubkey must be set in order to verify
+     * the received certificate.
+     */
+    Sv2HandshakeState(CKey&& static_key,
+                      XOnlyPubKey authority_pubkey) : m_static_key{static_key},
+                                                        m_authority_pubkey{authority_pubkey}
+    {
+        m_our_static_ellswift_pk = static_key.EllSwiftCreate(MakeByteSpan(GetRandHash()));
+    };
+
+    /*
+     * If we are the responder, the certificate must be set
+     */
+    Sv2HandshakeState(CKey&& static_key,
+                      Sv2SignatureNoiseMessage&& certificate) : m_static_key{static_key},
+                                                                m_certificate{certificate}
+    {
+        m_our_static_ellswift_pk = static_key.EllSwiftCreate(MakeByteSpan(GetRandHash()));
+    };
+
+    /** Handshake step 1 for initiator: -> e */
+    void WriteMsgEphemeralPK(Span<std::byte> msg);
+    /** Handshake step 1 for responder: -> e */
+    void ReadMsgEphemeralPK(Span<std::byte> msg);
+    /** During handshake step 2, put our ephmeral key, static key
+     * and certificate in the buffer: <- e, ee, s, es, SIGNATURE_NOISE_MESSAGE
+     */
+    void WriteMsgES(Span<std::byte> msg);
+    /** During handshake step 2, read the remote ephmeral key, static key
+     * and certificate. Verify their certificate.
+     * <- e, ee, s, es, SIGNATURE_NOISE_MESSAGE
+     */
+    [[nodiscard]] bool ReadMsgES(Span<std::byte> msg);
+
+    std::array<Sv2CipherState, 2> SplitSymmetricState();
+    uint256 GetHashOutput();
+
+    void SetEphemeralKey(CKey&& key);
+
+private:
+    /** Our static key (s) */
+    CKey m_static_key;
+    /** EllSwift encoded static key, for optimized ECDH */
+    EllSwiftPubKey m_our_static_ellswift_pk;
```

Comment:
styling nit: Consider dropping the `_our` part from the variable's name to be more consistent with the other names and be closer to the Noise protocol naming convention `s` (implicitly our) and `rs` (explicitly remote). Same for the rest of the names below.

Reply:
Done